tests/kafka.rs

```diff
@@ -0,0 +1,80 @@
+use futures::{StreamExt, TryStreamExt};
+use rdkafka::{
+    config::ClientConfig,
+    consumer::{stream_consumer::StreamConsumer, Consumer},
+    producer::{FutureProducer, FutureRecord},
+    Message,
+};
+use std::time::Duration;
+
+use testcontainers::{clients, core::RunArgs, images::kafka};
+
+#[tokio::test]
+async fn test_produce_and_consume_messages() {
+    let docker = clients::Cli::default();
+    let kafka_node = docker.run_with_args(
+        kafka::Kafka::default(),
+        RunArgs::default().with_mapped_port((kafka::KAFKA_PORT, kafka::KAFKA_PORT)),
```

Comment:
Like I commented on the other one, I believe this is still not necessary because the `KAFKA_ADVERTISED_LISTENERS` env variable is actually scoped within the container and there, the listening port is always the default one.

How docker maps this port to the host port should not matter!

Reply:
This means we can probably also get rid of the exported constant.

Reply:
I'd like to remove the port mapping as much as you do but cannot think of a way to get rid of it. Let me share this with you to better illustrate the pain point . Say I drop the port mapping so that we run a Kafka container as
```
let docker = clients::Cli::default();
docker.run(kafka::Kafka::default());
```
We then need to obtain the location of a boot strap server (so we can later create a Producer and a Consumer with that info) and this will look like
```
let bootstrap_servers = format!("localhost:{}", kafka_node.get_host_port(9092));  # only exposes 9092 by default
```
bootstrap_servers would probably look like `localhost:18291` that is, it will be assigned whatever port number bound to the container port 9092.

Here comes the problem. Given the exposed port `18291`, `KAFKA_ADVERTISED_LISTENERS` needs to specify `PLAINTEXT://localhost:18291,BROKER://localhost:9093` (BROKER has 9093 unlike my code actual code, but that does not change what we're talking about here). It's explained [here](https://www.confluent.io/blog/kafka-client-cannot-connect-to-broker-on-aws-on-docker-etc/#scenario-4) in more details, but the idea is we need to specify the host port  bound to 9092 . If, instead, `KAFKA_ADVERTISED_LISTENERS` specified `PLAINTEXT://localhost:9092,BROKER://localhost:9093` then the Producer in the test code would fail to publish messages due to a timeout error. In our case, a Kafka is running within a container and the test code is running on a bare metal. For the test code, including the Producer and the Consumer, to be able to talk to the Kafka broker running in the container, `KAFKA_ADVERTISED_LISTENERS` needs to specify the exposed host port, but setting that port number for `KAFKA_ADVERTISED_LISTENERS` happens before `kafka_node.get_host_port(9092)` so we don't know what the port number to specify for `KAFKA_ADVERTISED_LISTENERS` while `kafka::Kafka::default()` runs.

To work around this problem, we seem to have to have a port mapping. This way, we know a priori what will be the exposed host port for 9092 and can specify that port number in `KAFKA_ADVERTISED_LISTENERS`. I hope I've made myself understood. Given that, do you think it's possible for us to remove the port mapping? Interestingly, Java and Python client libraries can obtain the exposed host port for a container port *while* the container is starting up.

Reply:
I've experimented and learned (a bit surprisingly) that the host port and container port need to be the same in order for the client running on a bare metal to talk to the Kafka broker running in a container.

If we started a container without a port mapping:
```
let docker = clients::Cli::default();
docker.run(kafka::Kafka::default());
```
and if we obtained the address of a boot strap server
```
let bootstrap_servers = format!("localhost:{}", kafka_node.get_host_port(9092));  # only exposes 9092 by default
```
`bootstrap_servers` would look like `localhost:18168` that is, it would be assigned an arbitrary port number bound to the container port 9092. Now, if we set `KAFKA_ADVERTISED_LISTENERS` to `PLAINTEXT://localhost:9092,BROKER://localhost:9093` (BROKER has 9093 unlike my code actual code, but that does not change what we're talking about), the Producer would not be able to talk to the Kafka broker any more (would issue a timeout error) and that's because 18168 is not specified in the port appearing in `KAFKA_ADVERTISED_LISTENERS` ([this](https://stackoverflow.com/questions/51514891/use-docker-compose-with-port-mapping-for-local-kafka-testing) can be related).

In sum, if there is a way to force the exposed host port number to be the same as the container port number, we can remove
```
with_mapped_port((kafka::KAFKA_PORT, kafka::KAFKA_PORT)),
```
FYI, this does not need to be `KAFKA_PORT`, any port number will work equally as long as we pass the same port numbers.

Reply:
Here is my conclusion after looking into this a bit more. With current `testcontainers-rs`, I don't think it's feasible to remove `with_mapped_port` when running a Kafka container. BTW, I'll stand corrected; what I said above regarding a port mapping has to have the same port numbers externally and internally is incorrect. With additional environment variables for Kafka, that restriction can be lifted.

### Problem 1 - Kafka container needs to expose additional port as it is starting up 
When a Kafka client, whether a Producer or a Consumer, running on a bare metal interacts with a Kafka broker running in a container, the Kafka container needs to expose additional port, besides the default 9092, as explained [here](https://www.confluent.io/blog/kafka-client-cannot-connect-to-broker-on-aws-on-docker-etc/#scenario-4). Out of curiosity, I ran an integration test from the Python's `testcontainers` for its Kafka image and saw it exposes port 9093  by default:
```
$ docker ps
CONTAINER ID   IMAGE                         COMMAND                  CREATED         STATUS         PORTS                               NAMES
3e18d59476ff   confluentinc/cp-kafka:5.4.3   "sh -c 'while [ ! -f…"   3 seconds ago   Up 2 seconds   9092/tcp, 0.0.0.0:55079->9093/tcp   hungry_proskuriakova
```
testcontainers for Python and Java can specify an additional port mapping *while the container is starting up* so they can specify the exposed host port for 9093 within `KAFKA_ADVERTISED_LISTENERS` (please see [here](https://github.com/testcontainers/testcontainers-java/blob/dabc99c73eb9832ed2971c42412dd56bbd4d15be/modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java#L56) and [here](https://github.com/testcontainers/testcontainers-python/blob/bc969c53680606d7fff6a2e8350f5a6f234fcd38/testcontainers/kafka.py#L46) for examples). With the Rust library, it does not look like we can do that because specifying port mapping [has been moved out](https://github.com/testcontainers/testcontainers-rs/commit/f95a5c22167b6b0093ec97479c6070fae134a498) from `Image` to `Docker`. As such, within `kafka::Kafka::default()`, we cannot write down the exposed port for 9093 within `KAFKA_ADVERTISED_LISTENERS` and we only get to know the exposed port for 9093 after `docker.run`.

### Problem 2 -  `KAFKA_ADVERTISED_LISTENERS` is not just used internally
Since the value specified for `KAFKA_ADVERTISED_LISTENERS` is used by the Kafka client running on a bare-metal even after the container has started up, it is not just internally used by the Kafka broker.

---
In the Rust library, if there were a way to specify the exposed port for 9093 within `Kafka::default` so that we could provide the following for `KAFKA_ADVERTISED_LISTENERS` (it's shown conceptually and obviously the syntax is incorrect),
```
PLAINTEXT://localhost:$(get_exposed_port(9093)),BROKER://localhost:9092
```
then we could remove the port mapping and the user could just run
```
let docker = Cli::default();
let container = docker.run(Kafka::default());
let bootstrap_servers = format!("localhost:{}", container.get_host_port(9093));
```

Kindly let me know what you think.

Reply:
Very interesting, thank you for digging into this topic!

I don't fully understand how the Python and Java implementations achieve this. The port mapping and starting of the container happen at the same time, hence we don't know what the port is being mapped to until after the container is started.

At the point, we have already supplied the startup command to the container so we can no longer update it.

Sounds like a hard problem to solve so I am tempted to defer it for now and merge what we already have.

Let me know another quick pass on the code and see if we can go ahead with this!

Reply:
Yeah, I think you put it succinctly. I'm good with what we have now as it should hopefully serve as a reasonable starting point.